### PR TITLE
Feat/231-explorer-infinite-loading-blocks

### LINKED ISSUE #231: Paginate blocks page
[ref](https://vegaprotocol.slack.com/archives/C8W4JQ27P/p1649346710632129?thread_ts=1649346418.169329&cid=C8W4JQ27P)

Paginate the blocks page to allow users to load more recent/older blocks. Possibly if easy add the "number of blocks since this page was loaded" as per desings.

### LINKED ISSUE #276: Move use fetch to react helpers
follow up (non-blocking): use the one from react helpers or block explorer

_Originally posted by @dexturr in https://github.com/vegaprotocol/frontend-monorepo/pull/229#discussion_r853864209_

### PR DESCRIPTION
**Pull request by sam-keen**
Closes #231 
Closes #276

### CODE REVIEW

File: apps/explorer/src/app/routes/blocks/home/index.tsx [hunk 1]
```diff
@@ -1,29 +1,95 @@
...
+      if (lastBlockHeightLoaded) {
+        setNextBlockHeightToLoad(lastBlockHeightLoaded - 1);
+        setHasMoreBlocks(lastBlockHeightLoaded > 1);
+      }
+
+      setBlocksData([...blocksData, ...blockMetas]);
+    }
+
+    if (tmError) {
+      setError(tmError);
```
Comment (mattrussell36) [inline apps/explorer/src/app/routes/blocks/home/index.tsx:70]: question: Will we only catch an error if the initial request errors? Perhaps everything from the `await refetch` line should be wrapped in a try catch to pick up subsequent errors

File: apps/token/src/routes/governance/proposal/proposal-container.tsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import { useTranslation } from 'react-i18next';
 import { useParams } from 'react-router-dom';
 
 import { SplashLoader } from '../../../components/splash-loader';
-import useFetch from '../../../hooks/use-fetch';
+import { useFetch } from '@vegaprotocol/react-helpers';
```
Comment (mattrussell36) [inline apps/token/src/routes/governance/proposal/proposal-container.tsx:8]: Just want to verify that useFetch in token works in the same way as the version in react helpers and we can do a straight swap? If so 🙌

File: libs/react-helpers/src/hooks/use-fetch.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    if (initialFetch) {
+      fetchCallback();
+    }
+  }, [fetchCallback, initialFetch, url]);
+
+  useEffect(() => {
+    // Use the cleanup function for avoiding a possibly...
+    // ...state update after the component was unmounted
+    return () => {
+      console.log('unmount');
```
Comment (mattrussell36) [inline libs/react-helpers/src/hooks/use-fetch.ts:109]: chore (blocking):
```suggestion
```

File: libs/react-helpers/src/hooks/use-fetch.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      cancelRequest.current = true;
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (mattrussell36) [inline libs/react-helpers/src/hooks/use-fetch.ts:112]: chore (blocking): Dont think we need to disable this line as we are referencing a ref
```suggestion
```

File: apps/explorer/src/app/routes/blocks/home/index.tsx [hunk 5]
```diff
@@ -1,29 +1,95 @@
...
+import { BlocksInfiniteList } from '../../../components/blocks/blocks-infinite-list';
 import { JumpToBlock } from '../../../components/jump-to-block';
-import { t } from '@vegaprotocol/react-helpers';
+import { t, useFetch } from '@vegaprotocol/react-helpers';
+
+// This constant should only be changed if Tendermint API changes the max blocks returned
+const TM_BLOCKS_PER_REQUEST = 20;
 
 const Blocks = () => {
+  const [hasMoreBlocks, setHasMoreBlocks] = useState(true);
```
Comment (mattrussell36) [inline apps/explorer/src/app/routes/blocks/home/index.tsx:17]: suggestion (non blocking): Quite a lot of use states here, might be nice to refactor into a reducer or put everything into a single object.

**Review by mattrussell36 — CHANGES_REQUESTED** [review]
Looks good couple questions for you and some small things to fix.

File: apps/token/src/routes/governance/proposal/proposal-container.tsx (hunk 2 above)
Comment (sam-keen) [reply apps/token/src/routes/governance/proposal/proposal-container.tsx:8]: Yes, the changes to the Explorer version were around passing optional parameters to refetch. I tested it manually on Token before commiting, all's well

File: apps/explorer/src/app/routes/blocks/home/index.tsx (hunk 1 above)
Comment (sam-keen) [reply apps/explorer/src/app/routes/blocks/home/index.tsx:70]: Great point, thanks

File: apps/explorer/src/app/routes/blocks/home/index.tsx (hunk 5 above)
Comment (sam-keen) [reply apps/explorer/src/app/routes/blocks/home/index.tsx:17]: Good plan

File: apps/explorer/src/app/components/blocks/blocks-infinite-list.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  if (!blocks) {
+    return <div>No items</div>;
+  }
+
+  // If there are more items to be loaded then add an extra row to hold a loading indicator.
+  const itemCount = hasMoreBlocks ? blocks.length + 1 : blocks.length;
+
+  // Pass an empty callback to InfiniteLoader in case it asks us to load more than once.
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  const loadMoreItems = areBlocksLoading ? () => {} : loadMoreBlocks;
```
Comment (dexturr) [inline apps/explorer/src/app/components/blocks/blocks-infinite-list.tsx:39]:
```suggestion
  const loadMoreItems = areBlocksLoading ? NOOP : loadMoreBlocks;
```

and at the top of the file in module scope `const NOOP = () => {}`

File: apps/explorer/src/app/components/blocks/blocks-infinite-list.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import React from 'react';
```
Comment (dexturr) [inline apps/explorer/src/app/components/blocks/blocks-infinite-list.tsx:1]: chore (blocking): tests ;)

File: apps/explorer/src/app/components/blocks/blocks-refetch.tsx [hunk 8]
```diff
@@ -1,3 +1,5 @@
+import { useState, useEffect } from 'react';
```
Comment (dexturr) [inline apps/explorer/src/app/components/blocks/blocks-refetch.tsx:1]: chore (blocking): Jest tests please 👍

File: apps/explorer-e2e/src/support/pages/blocks-page.ts [hunk 9]
```diff
@@ -16,6 +16,8 @@ export default class BlocksPage extends BasePage {
   jumpToBlockSubmit = 'go-submit';
 
   validateBlocksPageDisplayed() {
+    // eslint-disable-next-line cypress/no-unnecessary-waiting
+    cy.wait(1000); // Wait for blocks to load if there are any
```
Comment (dexturr) [inline apps/explorer-e2e/src/support/pages/blocks-page.ts:20]:
```suggestion
```

**Pull request merged**